File: core/src/txn/engine.ts
/**
 * Upgrade transaction engine — pure state machine over TxnEffects +
 * HostAdapter (sim-first: zero direct IO/time/randomness; see effects.ts).
 *
 * Invariants (each is a harness tooth):
 *  - WAL: intent is journaled+fsync'd BEFORE the action it names.
 *  - Crash anywhere -> recover() lands on stable-running or completes the
 *    transition, decided by journal replay. Never dual-run, never bricked.
 *  - Promote only after the caller-supplied predicate evaluation passed.
 *  - Readback is bound to a NEW incarnation: the startId probed before
 *    handover is journaled, and evidence carrying it again is refused.
 *  - Rollback is always available until promote; its reason is journaled.
 */
import type { HostAdapter, ProcessEvidence } from "../lifecycle/hostAdapter.ts";
import type { TxnEffects } from "./effects.ts";
import type { JournalEntry, TxnPhase } from "./state.ts";
import { STATE_FORMAT_VERSION } from "./state.ts";
import type { Clock } from "../clock.ts";
import { HostCallTimeout, HostCallUncertain, DEFAULT_HOST_CALL_BUDGET_MS } from "./hostCallBudget.ts";


export interface EngineDeps {
  effects: TxnEffects;
  host: HostAdapter;
  clock: Clock;
  /**
   * Evaluate convergence for the freshly started experiment process.
   * Returns null when converged; otherwise a human-readable refusal that
   * becomes the rollback reason. (Full ConvergenceReport wiring lands with
   * converge/; the engine only cares about pass/fail + reason.)
   */
  evaluatePredicates: (evidence: ProcessEvidence, targetVersion: string) => Promise<string | null>;
  /**
   * Budget for a single host call (quiesce/stop/start/probe).
   *
   * A host that HANGS is worse than one that crashes: nothing is journaled,
   * the lock stays held by a process that is still alive (so stale-lock
   * takeover does not apply), and every later attempt queues behind it
   * forever. That is the "wedged half-way" failure, and it is the one an
   * updater is least able to explain afterwards.
   *
   * Default 120s: long enough for a real service to drain sessions on a busy
   * machine, short enough that a wedge is reported the same day it happens.
   */
  hostCallBudgetMs?: number;
}

export type EngineOutcome =
  | { result: "promoted"; version: string }
  | { result: "rolled-back"; reason: string }
  | { result: "up-to-date" };

export class UpgradeEngine {
  private readonly deps: EngineDeps;
  private seq = 0;

  constructor(deps: EngineDeps) {
    const budget = deps.hostCallBudgetMs ?? DEFAULT_HOST_CALL_BUDGET_MS;
    if (!Number.isSafeInteger(budget) || budget <= 0 || budget > 2_147_483_647) throw new Error("invalid host call budget");
    this.deps = deps;
  }

  private async journal(intent: TxnPhase, detail: Record<string, string> = {}): Promise<void> {
    const entry: JournalEntry = {
      seq: this.seq++,
      timestampMs: this.deps.clock.nowMs(),
      intent,
      detail: { ...detail, formatVersion: String(STATE_FORMAT_VERSION) },
    };
    await this.deps.effects.journal.appendAndSync(entry);
  }

  /**
   * Replay the journal and finish or undo whatever was in flight.
   * Must be called before upgrade() on every process start.
   */
  async recover(): Promise<void> {
    if (this.deps.host.fence) await this.withBudget("fence", () => this.deps.host.fence!());
    const entries = await this.deps.effects.journal.readAll();
    const last = entries.at(-1);
    this.seq = (last?.seq ?? -1) + 1;
    if (!last) return; // fresh install, stable running

    const version = (await this.deps.effects.slots.slotVersions()).experiment;
    switch (last.intent) {
      case "idle":
        return;
      case "promoted":
        // WAL redo: the intent is durable but its action may not have run
        // (crash in the after-journal window). Completing it is idempotent —
        // promoteExperiment on an already-promoted world is a no-op because
        // the experiment slot is empty.
        await this.deps.effects.slots.promoteExperiment();
        // resume is part of the terminal action too. A crash after promote()
        // but before resume() used to leave the service alive and its hosted
        // work permanently parked; DST found this exact effect boundary.
        await this.withBudget("resume", () => this.deps.host.resume());
        return;
      case "rolled-back":
        // The terminal journal entry is WAL intent, not proof that the host
        // restore ran. Redo the whole idempotent rollback action: a crash
        // immediately after journaling `rolled-back` may still have the
        // experiment process live and workloads parked.
        await this.withBudget("stop", () => this.deps.host.stop("experiment"));
        await this.withBudget("start", () => this.deps.host.start("stable"));
        await this.withBudget("resume", () => this.deps.host.resume());
        await this.deps.effects.slots.clearExperiment();
        return;
      case "staged":
        // Download completed but handover never started: cheap undo.
        await this.rollbackTo("crash before handover", { skipHostRestart: true });
        return;
      case "handing-over":
      case "running-experiment":
      case "readback": {
        // Runner death before the commit intent always undoes the attempt.
        // A healthy candidate does not authorize a successor to commit it.
        await this.rollbackTo(`crash during ${last.intent}` + (version ? ` (experiment ${version})` : ""));
        return;
      }
      default: {
        // Unknown intent => journal written by a NEWER core. Fail closed.
        throw new Error(
          `journal intent ${JSON.stringify(last.intent)} is not understood by this core (state format newer than binary); refusing to act`,
        );
      }
    }
  }

  /** Full transactional upgrade to targetVersion via a pre-verified artifact. */
  async upgrade(target: { version: string; bytesRef: string }): Promise<EngineOutcome> {
    const versions = await this.deps.effects.slots.slotVersions();
    if (versions.stable === target.version) return { result: "up-to-date" };

    // Record the incarnation this upgrade replaces, before anything on disk
    // changes. A probe that FAILS means nothing is live to compare against (a
    // stopped service is still upgradable); a probe that WEDGES is not
    // information and is let out, exactly as for the readback probe below.
    let prior: ProcessEvidence | null = null;
    try {
      prior = await this.withBudget("healthProbe", () => this.deps.host.healthProbe());
    } catch (err) {
      if (err instanceof HostCallUncertain) throw err;
      prior = null;
    }

    await this.journal("staged", { version: target.version });
    await this.deps.effects.slots.stageExperiment(target);

    // The external runner survives normal service replacement.
    await this.journal("handing-over", { version: target.version, ...(prior ? { priorStartId: prior.startId } : {}) });
    await this.withBudget("quiesce", () => this.deps.host.quiesce());
    await this.withBudget("stop", () => this.deps.host.stop("stable"));
    await this.withBudget("start", () => this.deps.host.start("experiment"));

    await this.journal("running-experiment", { version: target.version });
    let evidence: ProcessEvidence;
    try {
      evidence = await this.withBudget("healthProbe", () => this.deps.host.healthProbe());
    } catch (err) {
      // A probe that FAILED is information: the host answered "not healthy",
      // so rolling back (which calls stop/start/resume) is sound. A probe that
      // WEDGED is not information -- we do not know what the host is doing, and
      // issuing more host calls to a host that never answered the last one is
      // how a stuck upgrade becomes two live incarnations. Let it out; the
      // journal keeps the in-flight phase and the next start resolves it from
      // evidence.
      if (err instanceof HostCallUncertain) throw err;
      return this.rollbackOutcome(`experiment probe failed: ${(err as Error).message}`);
    }

    await this.journal("readback", { version: target.version });
    // The same startId after stop/start means the old process answered: it
    // was never stopped, or the probe served cached evidence. A version
    // string alone cannot tell those apart; the incarnation identity can.
    if (prior !== null && evidence.startId === prior.startId) {
      return this.rollbackOutcome(
        `live process is still the pre-upgrade incarnation (startId ${evidence.startId}); the old service was not replaced`,
      );
    }
    const refusal = await this.withBudget("readback", () => this.deps.evaluatePredicates(evidence, target.version));
    if (refusal !== null) {
      return this.rollbackOutcome(`predicates refused: ${refusal}`);
    }

    await this.journal("promoted", { version: target.version });
    await this.deps.effects.slots.promoteExperiment();
    await this.withBudget("resume", () => this.deps.host.resume());
    return { result: "promoted", version: target.version };
  }

  /**
   * Bound a host call. On expiry we do NOT pretend the call failed cleanly:
   * a pending promise cannot be cancelled in JS, so the host may still be
   * mid-operation. The transaction gives up instead of issuing further host
   * calls it cannot reason about, and the journal records why. Recovery is
   * the next process start, which replays the durable commit or restores stable.
   */
  private async withBudget<T>(label: string, call: () => Promise<T>): Promise<T> {
    const budgetMs = this.deps.hostCallBudgetMs ?? DEFAULT_HOST_CALL_BUDGET_MS;
    let cancel: (() => void) | undefined;
    try {
      return await Promise.race([
        call(),
        new Promise<never>((_resolve, reject) => {
          cancel = this.deps.clock.after(budgetMs, () => {
            reject(new HostCallTimeout(label, budgetMs));
          });
        }),
      ]);
    } finally {
      cancel?.();
    }
  }

  private async rollbackOutcome(reason: string): Promise<EngineOutcome> {
    await this.rollbackTo(reason);
    return { result: "rolled-back", reason };
  }

  private async rollbackTo(reason: string, opts: { skipHostRestart?: boolean } = {}): Promise<void> {
    await this.journal("rolled-back", { reason });
    if (!opts.skipHostRestart) {
      // Stop whatever may be running (either slot), restore stable, resume.
      await this.withBudget("stop", () => this.deps.host.stop("experiment"));
      await this.withBudget("start", () => this.deps.host.start("stable"));
      await this.withBudget("resume", () => this.deps.host.resume());
    }
    await this.deps.effects.slots.clearExperiment();
  }
}

File: src/engine.rs
//! The state machine has no direct disk/process/time dependencies. All effects
//! are explicit so every effect boundary can be crashed by the Rust harness.
use crate::{
    Error, Result,
    error::invalid,
    host::Host,
    state::{Evidence, JournalEntry, Phase, Slot},
    storage::Effects,
};
use async_trait::async_trait;
use std::{collections::BTreeMap, future::Future, path::Path, time::Duration};

#[async_trait]
pub trait Predicates: Send + Sync {
    async fn evaluate(&self, evidence: &Evidence, target: &str) -> Result<Option<String>>;
}
pub struct VersionPredicate;
#[async_trait]
impl Predicates for VersionPredicate {
    async fn evaluate(&self, evidence: &Evidence, target: &str) -> Result<Option<String>> {
        evidence.validate()?;
        Ok((evidence.version != target).then(|| {
            format!(
                "live process reports {}, expected {target}",
                evidence.version
            )
        }))
    }
}
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EngineOutcome {
    Promoted(String),
    RolledBack(String),
    UpToDate,
}
pub struct Engine<'a> {
    effects: &'a dyn Effects,
    host: &'a dyn Host,
    predicates: &'a dyn Predicates,
    now: &'a (dyn Fn() -> u64 + Send + Sync),
    budget: Duration,
    seq: u64,
}
impl<'a> Engine<'a> {
    pub fn new(
        effects: &'a dyn Effects,
        host: &'a dyn Host,
        predicates: &'a dyn Predicates,
        now: &'a (dyn Fn() -> u64 + Send + Sync),
        budget_ms: u64,
    ) -> Result<Self> {
        if budget_ms == 0 || budget_ms > 2_147_483_647 {
            return Err(invalid("invalid host call budget"));
        }
        Ok(Self {
            effects,
            host,
            predicates,
            now,
            budget: Duration::from_millis(budget_ms),
            seq: 0,
        })
    }
    async fn bounded<T>(&self, label: &str, call: impl Future<Output = Result<T>>) -> Result<T> {
        tokio::time::timeout(self.budget, call)
            .await
            .unwrap_or_else(|_| Err(Error::Uncertain(format!("HOST_CALL_TIMEOUT: {label}"))))
    }
    async fn journal(&mut self, intent: Phase, mut detail: BTreeMap<String, String>) -> Result<()> {
        detail.insert("formatVersion".into(), "1".into());
        let entry = JournalEntry {
            seq: self.seq,
            timestamp_ms: (self.now)(),
            intent,
            detail,
        };
        self.effects.append(&entry).await?;
        self.seq += 1;
        Ok(())
    }
    pub async fn recover(&mut self) -> Result<()> {
        // Validate persistent input before invoking any host effect, including fence.
        let entries = self.effects.read_journal().await?;
        for e in &entries {
            e.validate()?;
        }
        self.seq = entries.last().map_or(0, |e| e.seq + 1);
        self.bounded("fence", self.host.fence()).await?;
        let Some(last) = entries.last() else {
            return Ok(());
        };
        match last.intent {
            Phase::Idle => Ok(()),
            Phase::Promoted => {
                self.effects.promote().await?;
                self.bounded("resume", self.host.resume()).await
            }
            Phase::RolledBack => {
                self.bounded("stop", self.host.stop(Slot::Experiment))
                    .await?;
                self.bounded("start", self.host.start(Slot::Stable)).await?;
                self.bounded("resume", self.host.resume()).await?;
                self.effects.clear().await
            }
            Phase::Staged => self.rollback("crash before handover", true).await,
            Phase::HandingOver | Phase::RunningExperiment | Phase::Readback => {
                self.rollback(
                    &format!(
                        "crash during {}",
                        serde_json::to_value(last.intent)?
                            .as_str()
                            .unwrap_or("unknown")
                    ),
                    false,
                )
                .await
            }
        }
    }
    pub async fn upgrade(&mut self, version: &str, bytes: &Path) -> Result<EngineOutcome> {
        let versions = self.effects.versions().await?;
        let stable = versions
            .get("stable")
            .and_then(Option::as_deref)
            .ok_or_else(|| invalid("stable slot required before upgrade"))?;
        if stable == version {
            return Ok(EngineOutcome::UpToDate);
        }
        let prior = match self.bounded("probe", self.host.probe()).await {
            Ok(e) => Some(e),
            Err(e) if e.is_uncertain() => return Err(e),
            Err(_) => None,
        };
        let mut detail = BTreeMap::from([("version".into(), version.into())]);
        self.journal(Phase::Staged, detail.clone()).await?;
        self.effects.stage(version, bytes).await?;
        if let Some(e) = &prior {
            detail.insert("priorStartId".into(), e.start_id.clone());
        }
        self.journal(Phase::HandingOver, detail.clone()).await?;
        self.bounded("quiesce", self.host.quiesce()).await?;
        self.bounded("stop", self.host.stop(Slot::Stable)).await?;
        self.bounded("start", self.host.start(Slot::Experiment))
            .await?;
        detail.remove("priorStartId");
        self.journal(Phase::RunningExperiment, detail.clone())
            .await?;
        let evidence = match self.bounded("probe", self.host.probe()).await {
            Ok(e) => e,
            Err(e) if e.is_uncertain() => return Err(e),
            Err(e) => {
                return self
                    .rollback_outcome(&format!("experiment probe failed: {e}"))
                    .await;
            }
        };
        self.journal(Phase::Readback, detail.clone()).await?;
        if prior
            .as_ref()
            .is_some_and(|e| e.start_id == evidence.start_id)
        {
            return self
                .rollback_outcome("live process is still the pre-upgrade incarnation")
                .await;
        }
        if let Some(reason) = self
            .bounded("readback", self.predicates.evaluate(&evidence, version))
            .await?
        {
            return self
                .rollback_outcome(&format!("predicates refused: {reason}"))
                .await;
        }
        self.journal(Phase::Promoted, detail).await?;
        self.effects.promote().await?;
        self.bounded("resume", self.host.resume()).await?;
        Ok(EngineOutcome::Promoted(version.into()))
    }
    async fn rollback_outcome(&mut self, reason: &str) -> Result<EngineOutcome> {
        self.rollback(reason, false).await?;
        Ok(EngineOutcome::RolledBack(reason.into()))
    }
    async fn rollback(&mut self, reason: &str, skip_restart: bool) -> Result<()> {
        self.journal(
            Phase::RolledBack,
            BTreeMap::from([("reason".into(), reason.into())]),
        )
        .await?;
        if !skip_restart {
            self.bounded("stop", self.host.stop(Slot::Experiment))
                .await?;
            self.bounded("start", self.host.start(Slot::Stable)).await?;
            self.bounded("resume", self.host.resume()).await?;
        }
        self.effects.clear().await
    }
}
